akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala

```diff
@@ -124,12 +220,43 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
   /**
    * INTERNAL API
    */
-  @InternalApi private[akka] def updated(node: UniqueAddress, key: A, initial: B)(modify: B ⇒ B): ORMap[A, B] = {
-    val newValue = values.get(key) match {
-      case Some(old) ⇒ modify(old)
-      case _         ⇒ modify(initial)
+  @InternalApi private[akka] def updated(node: UniqueAddress, key: A, initial: B, valueDeltas: Boolean = false)(modify: B ⇒ B): ORMap[A, B] = {
+    val (oldValue, hasOldValue) = values.get(key) match {
+      case Some(old) ⇒ (old, true)
+      case _         ⇒ (initial, false)
+    }
+    // Optimization: for some types - like GSet, GCounter, PNCounter and ORSet  - that are delta based
+    // we can emit (and later merge) their deltas instead of full updates.
+    // However to avoid necessity of tombstones, the derived map type needs to support this
+    // with clearing the value (e.g. removing all elements if value is a set)
+    // before removing the key - like e.g. ORMultiMap does
+    def modifyWithDelta(value: DeltaReplicatedData) = {
+      val newValue = modify(value.resetDelta.asInstanceOf[B])
+      val newValueDelta = newValue.asInstanceOf[DeltaReplicatedData].delta
+      (newValue, newValueDelta)
     }
-    new ORMap(keys.add(node, key), values.updated(key, newValue))
+
+    // propagate 'put' where there was no old value - so effectively we put a new value
+    def gatherDelta(newValue: B, newValueDelta: Option[ReplicatedDelta]) = if (newValueDelta.isDefined && hasOldValue)
+      (newValue, UpdateDeltaOp(keys.resetDelta.add(node, key).delta.get, Map(key → newValueDelta.get)))
+    else
+      (newValue, PutDeltaOp(keys.resetDelta.add(node, key).delta.get, Map(key → newValue)))
+
+    val (newValue, modifiedDelta) = if (valueDeltas && oldValue.isInstanceOf[DeltaReplicatedData]) {
+      val (newValue, newValueDelta) = modifyWithDelta(oldValue.asInstanceOf[DeltaReplicatedData])
+      gatherDelta(newValue, newValueDelta)
```

Comment:
the `gatherDelta` reads nicely but to avoid the allocation of tuple instance I think it should be inlined (only used at one place), same with `modifyWithDelta`